specs/derivation.md

```diff
@@ -872,106 +933,118 @@ Queue) to start (L1 Traversal).
-  derived from L1. These blocks either come from sequencing (in sequencer mode) or from "unsafe sync" to the sequencer
-  (in validator mode).
-
-When resetting the L1 head, we need to rollback the safe head such that the L1 origin of the new safe head is a
-canonical L1 block (i.e. an the new L1 head, or one of its ancestors). We achieved this by walking back the L2 chain
-(starting from the current safe head) until we find such an L2 block. While doing this, we must take care not to walk
-past the [L2 genesis][g-l2-genesis] or L1 genesis.
-
-The unsafe head does not necessarily need to be reset, as long as its L1 origin is *plausible*. The L1 origin of the
-unsafe head is considered plausible as long as it is in the canonical L1 chain or is ahead (higher number) than the head
-of the L1 chain. When we determine that this is no longer the case, we reset the unsafe head to be equal to the safe
+- The [safe L2 head][g-safe-l2-head]: everything up to and including this block can be fully derived from the canonical
+  L1 chain.
+- The [unsafe L2 head][g-unsafe-l2-head]: blocks between the safe and unsafe heads are blocks that have not been derived
+  from L1. These blocks either come from sequencing (in sequencer mode) or from "unsafe sync" to the sequencer (in
+  validator mode).
+
+When resetting the engine queue because of an L1 re-org, we need to find new valid values for these two heads.
+
+- The new safe L2 head is the highest L2 block such that its [L1 origin][g-l1-origin] is a canonical L1 block (either
+  the new L1 head or one of its ancestors).
+- The new unsafe L2 head is the highest L2 block that is *plausible*, i.e. builds on top of the new safe L2 head.
+
+> **TODO** The above description does not match the implementation: we always reset to the highest *epoch* matching a
+> canonical L1 block, and set the its first L2 block as new L2 safe head. Would love to explore if this can be changed
+> to adopt the simpler definition.
+
+To do this, we will need to walk the L2 chain backwards, starting at the old safe L2 head, until we find the new safe L2
 head.
 
-> **TODO** Don't we always need to discard the unsafe head when there is a L1 re-org, because the unsafe head's origin
-> builds on L1 blocks that have been re-orged away?
+In doing so we need to take care not to re-org past the [L2 genesis][g-l2-genesis].
+
+> **TODO** This was in the spec:
+>
+> In practice, we'll pick an already-finalized L1 block as L2 inception point to preclude the possibility of a re-org
+> past genesis, at the cost of a few empty blocks at the start of the L2 chain.
 >
-> I'm guessing maybe we received some unsafe blocks that build upon the re-orged L2, which we accept without relating
-> them back to the safe head?
+> This makes sense, but is in conflict with how the [L2 chain inception][g-inception] is currently determined,
+> which is via the L2 output oracle deployment & upgrades.
```

Comment:
I'd like clarity on this — what's our plan for deployment? @tynes @smartcontracts